src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/VisitService.kt

```diff
@@ -0,0 +1,169 @@
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.data.CancelVisitRequest
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.data.CreateVisitRequest
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.data.CreateVisitResponse
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.EventStatus
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.EventStatus.Companion.SCHEDULED_APPROVED
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.OffenderBooking
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.OffenderVisitBalanceAdjustment
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.Visit
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.VisitOrderAdjustmentReason
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.VisitStatus
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.VisitType
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.VisitVisitor
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.AgencyLocationRepository
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.OffenderBookingRepository
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.OffenderVisitBalanceAdjustmentRepository
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.OffenderVisitBalanceRepository
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.ReferenceCodeRepository
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.VisitRepository
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.VisitVisitorRepository
+import java.time.LocalDate
+import java.time.LocalDateTime
+import java.util.function.Supplier
+
+@Service
+@Transactional
+class VisitService(
+  private val visitRepository: VisitRepository,
+  private val visitVisitorRepository: VisitVisitorRepository,
+  private val offenderBookingRepository: OffenderBookingRepository,
+  private val offenderVisitBalanceRepository: OffenderVisitBalanceRepository,
+  private val offenderVisitBalanceAdjustmentRepository: OffenderVisitBalanceAdjustmentRepository,
+  private val eventStatusRepository: ReferenceCodeRepository<EventStatus>,
+  private val visitTypeRepository: ReferenceCodeRepository<VisitType>,
+  private val visitStatusRepository: ReferenceCodeRepository<VisitStatus>,
+  private val agencyRepository: AgencyLocationRepository,
+  private val telemetryClient: TelemetryClient,
+) {
+  companion object {
+    private val log = LoggerFactory.getLogger(this::class.java)
+  }
+
+  fun createVisit(visitDto: CreateVisitRequest): CreateVisitResponse {
+
+    val offenderBooking = offenderBookingRepository.findByOffenderNomsIdAndActive(visitDto.offenderNo, true)
+      .orElseThrow(PrisonerNotFoundException(visitDto.offenderNo))
+
+    createBalance(visitDto, offenderBooking)
+
+    val visit = visitRepository.save(mapVisitModel(visitDto, offenderBooking))
+
+    addVisitors(visit, offenderBooking, visitDto)
```

Comment:
Probably worth looking at the mapping to understand why we need to do this as a separate step. Feels like we should be able to just create a Visit with a list of Visitors and save in one go rather then using a visitVistorRepository

Reply:
This looks tricky, im getting : 
Not-null property references a transient value - transient instance must be saved before current operation,
i.e. the Visitor's 'visit' property contains the raw transient Visit